Expose /readyz & /livez in kube-controller-manager

nojnhuh opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR supersedes #125276 by @nayihz: 

> Adds a proper liveness and readiness endpoints to the kube-controller-manager.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->
Fixes #118158

#### Special notes for your reviewer:

I left the history of this branch messy for now in case that's useful to review the delta from the previous PR. I think the only outstanding comment there was https://github.com/kubernetes/kubernetes/pull/125276#discussion_r1626581970 which I've addressed as a separate commit here. I plan to rebase/squash before merging.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kube-controller-manager: exposed the /livez and /readz endpoints for health checks that are in compliance with https://kubernetes.io/docs/reference/using-api/health-checks/#api-endpoints-for-health
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

nayihz committed:
Expose /readyz & /livez in kube-controller-manager

Co-authored-by: Jon Huhn <nojnhuh@users.noreply.github.com>

neolit123 reviewed (commented):
thanks for the PR!

let's get the commit prepared properly, first. please squash it to 1 commit, explain the feature in the commit message briefly and add a `Co-authored by nayihz`  at the bottom to provide the attribution for the original contributor.

as is, CI will block the merge commit and EOD we want a single commit for this, pretty much.

neolit123 commented:
/release-note-edit

```release-note
kube-controller-manager: exposed the /livez and /readz endpoints for health checks that are in compliance with https://kubernetes.io/docs/reference/using-api/health-checks/#api-endpoints-for-health
```

neolit123 commented on test/integration/controllermanager/serving/healthcheck_test.go:
```diff
@@ -0,0 +1,236 @@
+	brokenConfigStr := strings.ReplaceAll(configStr, "127.0.0.1", "127.0.0.2")
+	brokenConfig, err := os.CreateTemp("", "kubeconfig")
+	if err != nil {
+		t.Fatalf("Failed to create config file: %v", err)
```
```suggestion
		t.Fatalf("Failed to create kubeconfig file: %v", err)
```

neolit123 commented on test/integration/controllermanager/serving/healthcheck_test.go:
```diff
@@ -0,0 +1,236 @@
+	url := fmt.Sprintf("https://%s", secureInfo.Listener.Addr().String())
+	url = strings.ReplaceAll(url, "[::]", "127.0.0.1") // switch to IPv4 because the self-signed cert does not support [::]
+
+	// read self-signed server cert disk
```
```suggestion
	// read the self-signed server cert from disk
```

neolit123 commented on test/integration/controllermanager/serving/main_test.go:
```diff
@@ -0,0 +1,27 @@
+/*
+Copyright 2024 The Kubernetes Authors.
```
same

neolit123 reviewed (commented):
did a surface first pass.
code freeze for 1.35 is soon, but there is a change this can be merged by them.

neolit123 commented:
/priority important-longterm

nojnhuh commented:
Addressed comments and the integration and linter failures. Still looking into the e2e failure but interested to see if that fails the same way twice in a row.

neolit123 commented:
i think the rebase's not right. it pulled misc changes for various components and recent commits.

neolit123 commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -201,21 +202,36 @@ func Run(ctx context.Context, c *config.CompletedConfig) error {
 
-	// Setup any healthz checks we will want to use.
-	var checks []healthz.HealthChecker
+	// Setup any healthz healthzChecks we will want to use.
```
```suggestion
	// Setup any health checks we will want to use.
```

neolit123 commented on test/integration/controllermanager/serving/healthcheck_test.go:
```diff
@@ -0,0 +1,236 @@
+
+	apiserverConfig, err := os.CreateTemp("", "kubeconfig")
+	if err != nil {
+		t.Fatalf("Failed to create config file: %v", err)
```
kubeconfig

neolit123 commented on test/integration/controllermanager/serving/healthcheck_test.go:
```diff
@@ -0,0 +1,236 @@
+			}
+			req, err := http.NewRequest(http.MethodGet, base+tt.path, nil)
+			if err != nil {
+				t.Fatalf("failed to request: %v", err)
```
use consistent uppercase in t.Failed / t.Error e.g. `failed ..` -> `Failed ...`
more of that below.

nojnhuh commented on staging/src/k8s.io/controller-manager/app/serve.go:
```diff
@@ -55,18 +56,20 @@ func BuildHandlerChain(apiHandler http.Handler, authorizationInfo *apiserver.Aut
-	mux := mux.NewPathRecorderMux("controller-manager")
-	mux.Handle("/healthz", healthzHandler)
+func NewBaseHandler(c *componentbaseconfig.DebuggingConfiguration, checks, readyzChecks []healthz.HealthChecker) *mux.PathRecorderMux {
+	pathRecorderMux := mux.NewPathRecorderMux("controller-manager")
```
It seems like the linter doesn't complain. I'll change this back to `mux`.

neolit123 commented:
/remove-kind cleanup
/kind feature

/cc @siyuanfoundation 
you were a reviewer on the original PR, could you please review this new one?

neolit123 reviewed (commented):
thanks for the updates.
generally LGTM, but needs a review from #sig-api-machinery who own the KCM component. you can try asking in their slack, but unclear if a review can be done so close to code-freeze for 1.35.

nojnhuh commented on test/integration/serving/serving_test.go:
```diff
@@ -372,7 +372,7 @@ users:
 			wantErr:        false,
 			wantSecureCode: ptr.To(http.StatusOK),
-			wantPaths:      []string{"/configz", "/healthz", "/metrics"},
+			wantPaths:      []string{"/configz", "/healthz", "/livez", "/metrics", "/readyz"},
```
Rebased and this was the only conflict. The test now cares about the order of these paths so I had to go back to this order.

nojnhuh commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -201,21 +202,36 @@ func Run(ctx context.Context, c *config.CompletedConfig) error {
+		readyzChecks = append(readyzChecks, electionChecker)
 	}
-	healthzHandler := controllerhealthz.NewMutableHealthzHandler(checks...)
+	readyzChecks = append(readyzChecks, healthz.NewShutdownHealthz(ctx.Done()))
```
In my [latest push](https://github.com/kubernetes/kubernetes/compare/2b1cd576a0b27898eb6a865b602d8cffe1f21d78..c8c2e8ddf2cb4924a539e334b29996aa04a5ed41) I tried adding a background goroutine which calls `WaitForCacheSync` and then closes this channel.

nojnhuh replied:
I pushed a change to remove all the extra checks from `/healthz` so now it does the same thing as `/livez`, waiting for nothing before succeeding.

All of the initialization that used to be directed toward `/healthz` is now directed to `/readyz`. Is there anything we need to add to `/readyz` that wasn't already part of `/healthz`?

neolit123 reviewed (commented):
/retest
/remove-lifecycle rotten
